compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp

```diff
@@ -242,30 +258,61 @@ class ConvertFunc : public ConvertToLLVMPattern {
     assert(fnType.getNumInputs() == 0 && fnType.getNumResults() == 0);
 
     TypeConverter::SignatureConversion signatureConverter(/*numOrigInputs=*/0);
-    auto argMapping = getKernelArgMapping(funcOp);
-    // There may be dead symbols, we pick i32 pointer as default argument type.
-    SmallVector<Type, 8> llvmInputTypes(
-        argMapping.size(), LLVM::LLVMPointerType::get(rewriter.getContext()));
-    funcOp.walk([&](IREE::HAL::InterfaceBindingSubspanOp subspanOp) {
-      LLVM::LLVMPointerType llvmType;
-      if (auto memrefType = dyn_cast<BaseMemRefType>(subspanOp.getType())) {
-        unsigned addrSpace =
-            *getTypeConverter()->getMemRefAddressSpace(memrefType);
-        llvmType = LLVM::LLVMPointerType::get(rewriter.getContext(), addrSpace);
-      } else {
-        llvmType = LLVM::LLVMPointerType::get(rewriter.getContext());
-      }
-      llvmInputTypes[argMapping[subspanOp.getBinding()]] = llvmType;
-    });
-    // As a convention with HAL, push constants are appended as kernel arguments
-    // after all the binding inputs.
-    uint64_t numConstants = 0;
-    funcOp.walk([&](IREE::HAL::InterfaceConstantLoadOp constantOp) {
-      numConstants =
-          std::max(constantOp.getOrdinal().getZExtValue() + 1, numConstants);
-    });
-    llvmInputTypes.resize(argMapping.size() + numConstants,
-                          rewriter.getI32Type());
+    IREE::HAL::PipelineLayoutAttr layout;
+    llvm::SetVector<IREE::HAL::InterfaceBindingSubspanOp> subspans;
+    WalkResult subspanWalkRes =
+        funcOp.walk([&](IREE::HAL::InterfaceBindingSubspanOp subspanOp) {
+          auto thisLayout = subspanOp.getLayout();
+          subspans.insert(subspanOp);
+          if (!layout) {
+            layout = thisLayout;
+          }
+          if (layout != thisLayout) {
+            return WalkResult::interrupt();
+          }
+          return WalkResult::advance();
+        });
+    if (subspanWalkRes.wasInterrupted())
```

Comment:
This should be a verifier on ExecutableVariantOp as it can see the entry points and the module contents all at once. We need to uniformly support the same contract across all targets and it's easier to catch errors when things are produced than much later on when we're lowering here. Assume all constant ops and binding ops have the same layout, find the first one, and use that (will make this code much cleaner).

Reply:
Done

I take it there's currently nothing actually verifying people haven't mixed-and-matched pipeline layouts?